packages/backend/src/shared/utils.ts

```diff
@@ -28,3 +28,60 @@ export function isValidHttpUrl(receivedUrl: string): boolean {
 
   return true
 }
+
+export async function requestWithTimeout<T>(
+  request: () => Promise<T>,
+  timeoutMs: number
+): Promise<T> {
+  const timeout = async (): Promise<never> =>
+    new Promise((_, reject) =>
+      setTimeout(() => reject(new Error('Request timed out')), timeoutMs)
+    )
+
+  return Promise.race([request(), timeout()])
+}
+
+interface PollArgs<T> {
+  request(): Promise<T>
+  stopWhen?(result: T): boolean
+  pollingFrequencyMs: number
+  timeoutMs: number
+}
+
+export async function sleep(ms: number) {
+  return new Promise((r) => setTimeout(r, ms))
+}
+
+export async function poll<T>(args: PollArgs<T>): Promise<T> {
+  const {
+    request,
+    stopWhen = (response: T) => !!response,
+    timeoutMs,
+    pollingFrequencyMs
+  } = args
+
+  let elapsedTimeMs = 0
+  let response: T
+
+  do {
+    const requestStart = Date.now()
+
+    response = await requestWithTimeout(
+      () => request(),
+      timeoutMs - elapsedTimeMs
```

Comment:
`timeoutMs - elapsedTimeMs` essentially lets `timeoutMs` to be the maximum timeout for the polling period. i.e. we can break early without needing for the given request to complete before timing out